api/src/main/java/io/minio/MinioClient.java

```diff
@@ -200,76 +208,80 @@
-    amzHeaders.add("server-side-encryption-customer-key");
-    amzHeaders.add("server-side-encryption-customer-key-md5");
-    amzHeaders.add("website-redirect-location");
-    amzHeaders.add("storage-class");
+    AMZ_HEADERS.add("server-side-encryption");
+    AMZ_HEADERS.add("server-side-encryption-aws-kms-key-id");
+    AMZ_HEADERS.add("server-side-encryption-context");
+    AMZ_HEADERS.add("server-side-encryption-customer-algorithm");
+    AMZ_HEADERS.add("server-side-encryption-customer-key");
+    AMZ_HEADERS.add("server-side-encryption-customer-key-md5");
+    AMZ_HEADERS.add("website-redirect-location");
+    AMZ_HEADERS.add("storage-class");
   }
 
-  private static final Set<String> standardHeaders = new HashSet<>();
+  @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
+  private static final Set<String> STANDARD_HEADERS = new HashSet<>();
 
   static {
-    standardHeaders.add("content-type");
-    standardHeaders.add("cache-control");
-    standardHeaders.add("content-encoding");
-    standardHeaders.add("content-disposition");
-    standardHeaders.add("content-language");
-    standardHeaders.add("expires");
-    standardHeaders.add("range");
+    STANDARD_HEADERS.add("content-type");
+    STANDARD_HEADERS.add("cache-control");
+    STANDARD_HEADERS.add("content-encoding");
+    STANDARD_HEADERS.add("content-disposition");
+    STANDARD_HEADERS.add("content-language");
+    STANDARD_HEADERS.add("expires");
+    STANDARD_HEADERS.add("range");
   }
 
   private String userAgent = DEFAULT_USER_AGENT;
   private PrintWriter traceStream;
 
-  private HttpUrl baseUrl;
-  private String region;
-  private boolean isAwsHost;
+  private final HttpUrl baseUrl;
+  private final String region;
+  private final boolean isAwsHost;
   private boolean isAcceleratedHost;
   private boolean isDualStackHost;
   private boolean useVirtualStyle;
-  private String accessKey;
-  private String secretKey;
   private OkHttpClient httpClient;
 
+  private final CredentialsProvider credentialsProvider;
+
+  @SuppressWarnings("java:S107")
   private MinioClient(
       HttpUrl baseUrl,
       String region,
       boolean isAwsHost,
       boolean isAcceleratedHost,
       boolean isDualStackHost,
       boolean useVirtualStyle,
-      String accessKey,
-      String secretKey,
+      CredentialsProvider credentialsProvider,
       OkHttpClient httpClient) {
-    this.baseUrl = baseUrl;
+
     this.region = region;
+    this.baseUrl = baseUrl;
     this.isAwsHost = isAwsHost;
-    this.isAcceleratedHost = isAcceleratedHost;
+    this.httpClient = httpClient;
     this.isDualStackHost = isDualStackHost;
     this.useVirtualStyle = useVirtualStyle;
-    this.accessKey = accessKey;
-    this.secretKey = secretKey;
-    this.httpClient = httpClient;
+    this.isAcceleratedHost = isAcceleratedHost;
+    this.credentialsProvider =
+        Objects.requireNonNull(credentialsProvider, "Credentials provider can't be null");
```

Comment:
`Builder` should take care this check i.e. values from builder are already validated